Integrate local E2E test with dev environment

jonatansberg opened the pull request:
I've extended the default E2E test fixtures to detect the presence of containers initialized through `yarn dev`. When those containers are present we can reuse existing resources and use the lightweight dev backend image and gateway setup for E2E test isolation.

Since the `yarn dev` command won't start Tinybird by default I've split the default project into a separate analytics project and main admin project so that we can conditionally run those tests. In CI, the tests will still fail if Tinybird isn't available. 

Note: This PR includes the changes from #25551, which should be merged first.

jonatansberg committed:
Integrated local E2E test with dev env

refs https://linear.app/ghost/issue/BER-3077/improve-e2e-testing-integration

Detect the presence of containers initialized through yarn dev. Reuse existing
containers when available and use the lightweight dev backend image + gateway
setup for E2E test isolation.

jonatansberg committed:
Added conditional await for pageHitPromise to base public-page POM 

refs https://linear.app/ghost/issue/BER-3077/improve-e2e-testing-integration

Always waiting for the analytics request to succeed means any public test will
fail if Tinybird or the analytics service isn't running.  

Co-authored-by: Chris Raible <chris@ghost.org>

jonatansberg commented on e2e/helpers/environment/service-managers/dev-ghost-manager.ts:
```diff
@@ -0,0 +1,359 @@
+        }
+
+        debug('Migrations completed successfully');
+    }
```
Fixed

rob-ghost reviewed (approved):
I've run the following scenarios locally and all but one passed:

1. ✅ Running with `yarn dev:forward` - the running containers are used, analytics tests are skipped
2. ✅ Running standalone - e2e spins up the required containers (including tinybird)

I did find that running `yarn dev:analytics` which spins up tinybird and then running the tests they were still skipped even though the containers are available. I think this is intentional but I also think it would be possible to use them with a slight tweak to the service availability module. See [this comment](https://github.com/TryGhost/Ghost/pull/25552/files#r2577098442).

Assuming that might be out-of-scope, PR approved 👍🏻 (pending linting is fixed and the base branch is merged)

jonatansberg commented on e2e/helpers/environment/service-availability.ts:
```diff
@@ -0,0 +1,71 @@
+                name: [TINYBIRD.LOCAL_HOST],
+                status: ['running']
+            }
+        });
```
I tried fixing this. Now it works great for the `dev:analytics` case, but it's for some unknown reason a bit racy for the standalone E2E tests. I'll come back to this tomorrow.

ibalosh commented on e2e/helpers/environment/environment-factory.ts:
```diff
@@ -0,0 +1,60 @@
+ * Check if the dev environment (yarn dev:forward) is running.
+ * Detects by checking for the ghost_dev network and running MySQL container.
+ */
+export async function isDevEnvironmentAvailable(): Promise<boolean> {
```
Do you find this way sufficient to check whether dev environment is running locally?

ibalosh reviewed (changes requested):
I’ll take another pass tomorrow with a fresher head to fully understand the newly introduced managers, so please take this initial review with a bit of caution. I wanted to get some early feedback to you quickly.

Some of the concerns I had by looking at the implementation:

**Too much behaviour is hidden behind auto detection** 

* get environment manager silently decides dev and test (standalone) mode. If something is missconfigured, you silently fallback instead of failing loudly
* something like const useDev = process.env.E2E_DEV === 'true' maybe?
* also for analytics exclusion I would simply use [tag annotations](https://playwright.dev/docs/test-annotations#tag-tests)

there's a lot of coupling to docker internals, like if changing compose.dev.yaml naming, hard to find the error (I admit though our initial implementation has similar issue, but it's less complex in my opinion)

**Two environment managers instead of one**

Original already encapsulates the general idea of start, cleanup, stop etc, why create one more? Ideally, they should be two configurations, rather than two classes plus factory and auto detection?

We might want to add simpler setup of env to choose, simplify a bit env manager, and per config and then just:

```
export function getEnvironmentManager(): EnvironmentManager {
  const config = loadFromEnv();
  return new EnvironmentManager(config);
}
```

Thats some initial reactions I had by look today. I do realize that this can't be simple, but we should keep it still friendly if possible to new-commers to the e2e repo and keep it very well documented to be able to read it. At first glance, it seems more complex 

I also got errors on custom-views.tests when yarn dev:forward is running in one shell and I do yarn test in other

9larsons commented on e2e/helpers/environment/service-availability.ts:
```diff
@@ -0,0 +1,83 @@
+ * Returns true if Tinybird is not available AND we're not in CI.
+ * In CI, we want tests to fail if Tinybird isn't available.
+ */
+export async function shouldSkipAnalyticsTests(): Promise<boolean> {
```
I tend to agree on principle. While this is slick and a proactive setup, I'd rather force the engineer to be more prescriptive.

On the other hand, I think if you run `yarn test:e2e` it should run what we expect to run on CI. If that means we need to boot extra containers, then we should do that. And we shouldn't require the dev to know that they need to disable tinybird. What Jonathan is trying to do here is conserve resources/increase speed of running the tests. 

I don't think I like the default being a slimmer version of our tests, and this seems like a reasonable compromise. I'm ultimately on the fence.

9larsons reviewed (dismissed):
In a note above, plus a couple questions.

jonatansberg commented:
@ibalosh I missed your main comment, sorry. 

Your concerns around auto-detection and having two managers instead of one are definitely reasonable. I went for building a second manager and auto-detection rather than trying to iterate on the existing one, as doing so while maintaining support for both modes of operation became a mess. 

The auto-detection makes the default DX nicer, but we can definitely have an `e2e:forward` command or something to that effect and toggle based on environment variables if you feel more comfortable that way. 

Ultimately my goal would be for what's currently called the dev environment manager in this PR to supersede the current one. In that world, we would still have two different modes of running the tests, either against the local development environment or against pre-built assets, but we would be rid of most of the duplication. 

This would also allow running the tests locally in the same way as today, which would be slightly faster if you're only doing backend changes.

ibalosh commented:
hey @jonatansberg let's catchup on a call to talk more to move this one forward. Even though number of files is small-ish, the PR is a bit bigger, with multiple concepts added, on some I am sold, some not, so I would like to get clarity

jonatansberg committed:
Set up separate Playwright project for analytics

refs https://linear.app/ghost/issue/BER-3077/improve-e2e-testing-integration

Make it possible to run analytics tests separately from the rest of the E2E 
test suite. We should likely extend this further and split up the main project
into additional smaller projects and reorganize the tests accordingly.

cmraible reviewed (changes requested):
Sorry it's taken a while for me to take a good look at this; I've waited because I generally agree with @ibalosh and @9larsons feedback. That said, I do think we should work to enable this for the quicker inner loop when developing locally, so I'm motivated to find a solution here that we can all agree on. I'm also eager to fully adopt the `dev:forward` experience across the board and ultimately remove the need for the more complex, static Docker build.

One thing I'll add, which maybe (hopefully?) clears a path to a simpler solution: I don't see any reason to maintain two different "modes" of running the e2e tests, which is where I think a lot of this additional complexity is coming from. I'd much prefer to _replace_ the current method of running tests (both locally and in CI for parity) than to add another way to run them.

cmraible commented on e2e/tests/public/homepage.test.ts:
```diff
@@ -9,4 +9,15 @@ test.describe('Ghost Public - Homepage', () => {
+        await homePage.goto();
+
+        // Wait for the request to be fulfilled
+        await expect(response).resolves.toBeDefined();
```
Could we cleanup this syntax at all? 

One of our goals with the e2e test suite is to make the tests as readable as possible, such that someone non-technical can read and understand them. Keeping the assertion isolated in the Page Object abstracts away this more technical detail and makes the tests themselves read a lot nicer, and this feels like a regression in that sense.

It's also easy to forget that this is necessary when writing tests, and if we forget it our tests end up being significantly flakier

jonatansberg commented on e2e/helpers/environment/environment-factory.ts:
```diff
@@ -0,0 +1,48 @@
+    try {
+        const networks = await docker.listNetworks({
+            filters: {name: [DEV_ENVIRONMENT.networkName]}
+        });
```
The reason I kept it in there is to ensure that this works out of the box with the Playwright integration in VS Code or Cursor and for people who want to run the tests while developing things. I'm planning to move on to getting the standalone version of this to run next, which means we would be able to pull out the old setup. 

I think the network ought to be cleaned up when `yarn dev` exits since [we do call `docker compose down`](https://github.com/TryGhost/Ghost/blob/main/package.json#L136). I did have a check for if the MySQL container previously but I removed that in order to clean up the config.. I could add that back in tho 🤔

jonatansberg replied:
I updated this to check for all the required services rather than assuming they're there.

jonatansberg commented on e2e/helpers/pages/public/public-page.ts:
```diff
@@ -86,13 +86,11 @@ export class PublicPage extends BasePage {
 
     async goto(url?: string, options?: pageGotoOptions): Promise<void> {
         await this.enableAnalyticsRequests();
-        const pageHitPromise = this.pageHitRequestPromise();
```
I think I understand what you mean by flakiness now! It's the other analytics tests that become flaky 🤦

jonatansberg commented on e2e/playwright.config.mjs:
```diff
@@ -43,13 +43,22 @@ const config = {
+                viewport: {width: 1920, height: 1080}
+            },
+            dependencies: ['global-setup']
+        },
```
Done!

cmraible reviewed (approved):
Thanks for incorporating my feedback, LGTM 🚢

jonatansberg commented:
@9larsons let me know if you have any more feedback. I'd be happy to incorporate that as a follow-up. 

I'll go ahead and merge this so I can move on to getting the CI version of this done and ensuring we test against the latest public release in addition to main.
